sw/device/tests/hmac/dif_hmac_test.c

```diff
@@ -0,0 +1,116 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/dif/dif_hmac.h"
+
+#include "sw/device/lib/base/mmio.h"
+#include "sw/device/lib/common.h"
+#include "sw/device/lib/flash_ctrl.h"
+#include "sw/device/lib/uart.h"
+
+#define HMAC0_BASE_ADDR 0x40120000
+
+static const size_t kDataLen = 142;
+static const char kData[142] =
+    "Every one suspects himself of at least one of "
+    "the cardinal virtues, and this is mine: I am "
+    "one of the few honest people that I have ever "
+    "known";
+
+static const uint32_t hmac_key[8] = {0x243f6a88, 0x85a308d3, 0x13198a2e,
+                                     0x03707344, 0xa4093822, 0x299f31d0,
+                                     0x082efa98, 0xec4e6c89};
+
+static const uint32_t kExpectedShaDigest[8] = {
+    0xdc96c23d, 0xaf36e268, 0xcb68ff71, 0xe92f76e2,
+    0xb8a8379d, 0x426dc745, 0x19f5cff7, 0x4ec9c6d6};
+
+static const uint32_t kExpectedHmacDigest[8] = {
+    0xe4987b39, 0x3f83d390, 0xc2f3bbaf, 0x3195dbfa,
+    0x23fb480c, 0xb012ae5e, 0xf1394d28, 0x1940ceeb};
+
+void checkDigest(dif_hmac_digest_t digest, uint32_t *expected) {
+  for (int i = 0; i < 8; i++) {
+    if (expected[i] != digest.digest[i]) {
+      uart_send_str("Digest mismatch, expected: 0x");
+      uart_send_uint(expected[i], sizeof(uint32_t) * 8);
+      uart_send_str(" but got: 0x");
+      uart_send_uint(digest.digest[i], sizeof(uint32_t) * 8);
+      uart_send_str("\n\rFAIL!\n\r");
+      return;
+    }
+  }
+
+  uart_send_str("PASS!\n\r");
+}
+
+int main(int argc, char **argv) {
+  dif_hmac_config_t hmac_config;
+  dif_hmac_digest_t digest;
+  dif_hmac_t hmac;
+  char *data = (char *)kData;
+
+  uart_init(UART_BAUD_RATE);
+  uart_send_str("Running HMAC DIF test\r\n");
+
+  // Test SHA256
+  uart_send_str("Testing SHA256 mode...\r\n");
+
+  hmac_config.base_addr = mmio_region_from_addr(HMAC0_BASE_ADDR);
+  hmac_config.digest_swap = 1;
+  hmac_config.endian_swap = 1;
+  hmac_config.mode = DIF_HMAC_OP_SHA256;
+
+  dif_hmac_init(&hmac_config, &hmac);
+
+  while (data - kData < kDataLen) {
+    data += dif_hmac_push_fifo(&hmac, data, kDataLen - (data - kData));
+  }
+
+  dif_hmac_run(&hmac);
+
+  uart_send_str("Sent ");
+  uart_send_uint(dif_hmac_get_msg_len(&hmac), sizeof(uint32_t) * 8);
+  uart_send_str(" bytes.\r\n");
+
+  while (!dif_hmac_is_done(&hmac))
+    ;
+
+  dif_hmac_read_digest(&hmac, &digest);
+
+  checkDigest(digest, (uint32_t *)kExpectedShaDigest);
+
+  // Test HMAC
+  uart_send_str("Testing HMAC mode...\r\n");
+
+  data = (char *)kData;
+  hmac_config.mode = DIF_HMAC_OP_HMAC;
+
+  for (int i = 0; i < 8; ++i) {
+    hmac_config.key[i] = hmac_key[i];
+  }
+
+  dif_hmac_init(&hmac_config, &hmac);
+
+  while (data - kData < kDataLen) {
+    data += dif_hmac_push_fifo(&hmac, data, kDataLen - (data - kData));
+  }
+  for (volatile uint32_t i = 0; i < 10000; i++)
+    ;
```

Comment:
If you want to burn cycles, use one of the sleep functions in lib/runtime.

Reply:
Oh dear, not only did clang-tidy do some funky stuff here but I evidently left in some debugging code. I'll clean this up.